### ISSUE TITLE
[MNG-5682] Parent POMs not resolved in multi-module project

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6631]
**[Petr Novak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kek)** opened **[MNG-5682](https://issues.apache.org/jira/browse/MNG-5682?redirect=false)** and commented

I have multi-module project - I attach the similar simple project structure to this issue, to simulate the problem =>  !test-project.zip!.

The structure is:

```
A    - aggregating project, parent for "parent"
 |_parent  - parent for B and C
 |_B
 |_C
```

In reality we have more parents under A for diferent types of A-submodules, but now it does not matter.

When we run build under maven 2.2.1  - everything is OK, the reactor sorts the projects like  A, PARENT, B,C and build success.

When we run the same build under maven 3.x  (3.0.4, 3.1.1, 3.2.3 was tested), the build fails with following errors:

a>mvn clean
[INFO] Scanning for projects...
[ERROR] The build could not read 2 projects -> [Help 1]
[ERROR]
[ERROR]   The project a:b:[unknown-version] (\a\b\pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM: Could not find artifact a:parent:pom:0.0.1-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 6, column 10 -> [Help 2]
[ERROR]
[ERROR]   The project a:c:[unknown-version] (\a\c\pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM: Could not find artifact a:parent:pom:0.0.1-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 6, column 10 -> [Help 2]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException

There is no explicit "relativePath" set in project POMs.

---

**Affects:** 3.0.4, 3.1.1, 3.2.3

**Attachments:**
- [test-project.zip](https://issues.apache.org/jira/secure/attachment/12713282/test-project.zip) (_1.30 kB_)

2 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Petr Novak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kek)** commented

Thank you for your tip, but my problem has a little different project structure.  
The parent project is on the same level as submodules B and C. And all projects  A, PARENT, B, and C have the same groupId and version.
The maven build was executed from project "A"  - root of the structure.

My inheritance structure is:

```
A<--parent<--B
                 <--C
```

and submodules structure

```
A<>--parent
 <>--B
 <>--C
```

Today I found possible workaround/solution for this structure is to set "relativePath" on "B" and "C" to  "../parent/pom.xml".

In my opinion Maven 2.x  adds all projects to reactor and makes build in order "A", "parent", "B", "C"   - the "A" is build and installed in local repository first, so than could be resolved as parent for "parent", than "parent" is build and installed in local repository and could be resolved for "B" and "C" as parent project.
This corresponds to specification in http://maven.apache.org/ref/3-LATEST/maven-model/maven.html#class_parent
??Maven looks for the parent POM first in this location on the filesystem, then the local repository, and lastly in the remote repo. ...??

But  Maven 3.x  resolves the \<parent>  specification first, without build of "A", "parent"  - so it must throw Error, because  relativePath is not specified and projects "A","parent" are not build and installed into local repository  - so could not be resolved.

Is this backward incompatibility between Maven 2.x and 3.x feature or bug ?

**Comment by jira-importer** [comment]
**[John Wu](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=phantom_john)** commented

Got similar issue here.

My workaround is to build and install the parent pom first, then build the rest modules.

Ideally, if the parent pom is in the reactor build plan, the reactor should 1) build the parent pom prior to build its child modules, 2) pick up the parent pom from within the reactor and use it for the child modules.

**Comment by jira-importer** [comment]
**[Devin Reid](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=devman07)** commented

Ran in to this issue as well. This is documented behavior as indicated by [Maven 3.x compatiblity notes](https://cwiki.apache.org/confluence/display/MAVEN/Maven+3.x+Compatibility+Notes#Maven3.xCompatibilityNotes-ParentPOMResolution).

> Maven 3 no longer resolves parent POMs from the local project checkout unless the child POM's \<relativePath> element is accurate, whether explicitly given or using the default value. This improves consistency regarding the build result when building the child project in isolation and when performing a reactor build that includes the parent project. In Maven 2, building the child project in isolation could fail while the reactor build would succeed to resolve the parent.

The reasoning behind this doesn't make sense to me. This change impacts projects that want to "opt-out" of the of the relativePath functionality (i.e. set \<relativePath /> for parents in child poms) . If I set relativePath to be no path I certainly want the pom to be resolved from the reactor. More over it makes parent poms behavior as a dependency inconsistent with how every other dependency in Maven works.

The reason stated that behavior was inconsistent between isolated and reactor builds but I disagree. It was entirely consistent. Using the sample project attached to this issue as an example, if B has a dependency on C, it will fail to build in isolation if C hasn't been installed in exactly the same way it will fail to build if 'parent' hadn't been installed. That is consistent and the way Maven 2 worked. In Maven 3, if parent isn't already installed a reactor build fails, even thought the same would not be true if C were not already installed. Even worse, if parent is installed and I make changes to the working copy those changes are not picked up by dependents even though parent is in the reactor. Very inconsistent behavior.

I would hope that this new behavior for parent resolution could be rolled back in its entirety, but a less drastic change would be to allow resolving a parent from the reactor when 'relativePath' is set to null/empty. At least that would allow projects to 'opt-out' as before.
